jtransc-rt/src/java/util/Timer.java

```diff
@@ -1,588 +1,298 @@
-                try {
-                    task.run();
-                    taskCompletedNormally = true;
-                } finally {
-                    if (!taskCompletedNormally) {
-                        synchronized (this) {
-                            cancelled = true;
-                        }
-                    }
-                }
-            }
-        }
+     public int purge() {
+         int result = 0;
 
-        private void insertTask(TimerTask newTask) {
-            // callers are synchronized
-            tasks.insert(newTask);
-            this.notify();
-        }
+         synchronized(queue) {
+             for (int i = queue.size(); i > 0; i--) {
+                 if (queue.get(i).state == TimerTask.CANCELLED) {
+                     queue.quickRemove(i);
+                     result++;
+                 }
+             }
 
-        /**
-         * Cancels timer.
-         */
-        public synchronized void cancel() {
-            cancelled = true;
-            tasks.reset();
-            this.notify();
-        }
+             if (result != 0)
+                 queue.heapify();
+         }
 
-        public int purge() {
-            if (tasks.isEmpty()) {
-                return 0;
-            }
-            // callers are synchronized
-            tasks.deletedCancelledNumber = 0;
-            tasks.deleteIfCancelled();
-            return tasks.deletedCancelledNumber;
-        }
+         return result;
+     }
+}
 
-    }
+class TimerThread extends Thread {
+	boolean newTasksMayBeScheduled = true;
+	private TaskQueue queue;
+
+	TimerThread(TaskQueue queue) {
+		this.queue = queue;
+	}
+
+	public void run() {
+		try {
+			mainLoop();
+		} finally {
+			// Someone killed this Thread, behave as if Timer cancelled
+			synchronized (queue) {
+				newTasksMayBeScheduled = false;
+				queue.clear();  // Eliminate obsolete references
+			}
+		}
+	}
+
+	private void mainLoop() {
+		while (true) {
+			try {
+				if (!newTasksMayBeScheduled) {
+					return;
+				}
+
+				boolean taskFired = false;
+				TimerTask task = null;
+
+				synchronized (queue) {
+					if (!queue.isEmpty()) {
+						long currentTime, executionTime;
+						task = queue.getMin();
+						synchronized (task.lock) {
+							if (task.state == TimerTask.CANCELLED) {
+								queue.removeMin();
+								continue;  // No action required, poll queue again
+							}
+							currentTime = System.currentTimeMillis();
+							executionTime = task.nextExecutionTime;
+
+							if (taskFired = (executionTime <= currentTime)) {
+								if (task.period == 0) { // Non-repeating, remove
+									queue.removeMin();
+									task.state = TimerTask.EXECUTED;
+								} else { // Repeating task, reschedule
+									queue.rescheduleMin(task.period < 0 ? currentTime - task.period : executionTime + task.period);
+								}
+							}
+						}
+					}
+				}
+
+				if (!taskFired) { // Task hasn't yet fired; wait
+					Thread.sleep(50);
```

Comment:
This one is another target for a future optimization using synchronization primitives. I have to check the status of them right now, since the initial target of jtransc was single threaded environments probably they still need some work, so I can understand this if they are not fully working by now.